### PR TITLE
Sanity test fixups for AWS ec2 modules

### PR DESCRIPTION
**Pull request by tremble**
##### SUMMARY
Sanity test fixups for AWS ec2 modules

- standard python3 related boilierplate
- move type checking into arg parser so we don't document "type: str" when it's going to be cast to int
- deprecate a couple of unused options
- remove duplicate definition of region (ec2 doc/spec adds it)
- missing documentation of type / suboptions

##### ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request

##### COMPONENT NAME

       lib/ansible/modules/cloud/amazon/ec2.py
       lib/ansible/modules/cloud/amazon/ec2_ami.py
       lib/ansible/modules/cloud/amazon/ec2_ami_copy.py
       lib/ansible/modules/cloud/amazon/ec2_ami_info.py
       lib/ansible/modules/cloud/amazon/ec2_asg.py
       lib/ansible/modules/cloud/amazon/ec2_asg_info.py
       lib/ansible/modules/cloud/amazon/ec2_asg_lifecycle_hook.py
       lib/ansible/modules/cloud/amazon/ec2_customer_gateway.py
       lib/ansible/modules/cloud/amazon/ec2_customer_gateway_info.py
       lib/ansible/modules/cloud/amazon/ec2_eip.py
       lib/ansible/modules/cloud/amazon/ec2_eip_info.py
       lib/ansible/modules/cloud/amazon/ec2_eni.py
       lib/ansible/modules/cloud/amazon/ec2_eni_info.py
       lib/ansible/modules/cloud/amazon/ec2_group.py
       lib/ansible/modules/cloud/amazon/ec2_group_info.py
       lib/ansible/modules/cloud/amazon/ec2_key.py
       lib/ansible/modules/cloud/amazon/ec2_launch_template.py
       lib/ansible/modules/cloud/amazon/ec2_lc.py
       lib/ansible/modules/cloud/amazon/ec2_lc_find.py
       lib/ansible/modules/cloud/amazon/ec2_lc_info.py
       lib/ansible/modules/cloud/amazon/ec2_metadata_facts.py
       lib/ansible/modules/cloud/amazon/ec2_metric_alarm.py
       lib/ansible/modules/cloud/amazon/ec2_placement_group.py
       lib/ansible/modules/cloud/amazon/ec2_placement_group_info.py
       lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py
       lib/ansible/modules/cloud/amazon/ec2_snapshot.py
       lib/ansible/modules/cloud/amazon/ec2_snapshot_copy.py
       lib/ansible/modules/cloud/amazon/ec2_snapshot_info.py
       lib/ansible/modules/cloud/amazon/ec2_tag.py
       lib/ansible/modules/cloud/amazon/ec2_transit_gateway.py
       lib/ansible/modules/cloud/amazon/ec2_vol.py
       lib/ansible/modules/cloud/amazon/ec2_vol_info.py
       lib/ansible/modules/cloud/amazon/ec2_win_password.py

### CODE REVIEW

**Review by felixfontein — APPROVED** [review]
Looks good to me.

File: lib/ansible/modules/cloud/amazon/ec2_launch_template.py [hunk 1]
```diff
@@ -56,20 +60,26 @@
...
         description: >
           The virtual device name (ephemeralN). Instance store volumes are
           numbered starting from 0. An instance type with 2 available instance
           store volumes can specify mappings for ephemeral0 and ephemeral1. The
           number of available instance store volumes depends on the instance
           type. After you connect to the instance, you must mount the volume.
+        type: str
       ebs:
-        description: Parameters used to automatically set up EBS volumes when the instance is launched.
+        description: Parameters used to automatically set up EBS volumes when the instance is launchedself
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_launch_template.py:81]:
```suggestion
        description: Parameters used to automatically set up EBS volumes when the instance is launched
```

File: lib/ansible/modules/cloud/amazon/ec2_eip.py [hunk 2]
```diff
@@ -28,65 +28,73 @@
...
+        an Elastic IP if it is tagged with I(tag_name).
     version_added: "2.9"
+    type: str
   tag_value:
     description:
-      - Supplements tag_name but also checks that the value of the tag provided in tag_name matches tag_value.
-        matches the tag_value
-    default: 'no'
+      - Supplements I(tag_name) but also checks that the value of the tag provided in I(tag_name) matches I(tag_value).
+        matches the I(tag_value)
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_eip.py:85]:
```suggestion
```
You didn't add this, but this sentence fragment doesn't seem to belong here. Would you mind cleaning it up?

File: lib/ansible/modules/cloud/amazon/ec2_launch_template.py [hunk 3]
```diff
@@ -94,36 +105,45 @@
...
+        type: int
       threads_per_core:
         description: >
           The number of threads per CPU core. To disable Intel Hyper-Threading
           Technology for the instance, specify a value of 1. Otherwise, specify
           the default value of 2.
+        type: int
   credit_specification:
-    description: The credit option for CPU usage of the instance. Valid for T2 or T3 instances only.
+    description: The credit option for CPU usage of the instance. Valid for T2 or T3 instances onlyself.
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_launch_template.py:139]:
```suggestion
    description: The credit option for CPU usage of the instance. Valid for T2 or T3 instances only
```

File: lib/ansible/modules/cloud/amazon/ec2_launch_template.py [hunk 4]
```diff
@@ -140,69 +160,86 @@
...
           spot_instance_type:
             description: The request type to send.
             choices: [one-time, persistent]
-    type: dict
+            type: str
   instance_type:
     description: >
-      The instance type, such as I(c5.2xlarge). For a full list of instance types, see
+      The instance type, such as C(c5.2xlarge). For a full list of instance types, see
       http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_launch_template.py:217]:
```suggestion
      U(http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-types.html)
```

File: lib/ansible/modules/cloud/amazon/ec2_launch_template.py [hunk 5]
```diff
@@ -212,63 +249,71 @@
...
           EC2 automatically selects the IPv6 addresses from the subnet range.
           You can't use this option if specifying the I(ipv6_addresses) option.
+        type: int
       ipv6_addresses:
         description: >
           A list of one or more specific IPv6 addresses from the IPv6 CIDR
           block range of your subnet. You can't use this option if you're
           specifying the I(ipv6_address_count) option.
+        type: list
+        elements: int
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_launch_template.py:272]:
```suggestion
        elements: str
```

File: lib/ansible/modules/cloud/amazon/ec2_lc.py [hunk 6]
```diff
@@ -43,89 +35,117 @@
...
     description:
       - Determines whether the instance runs on single-tenant hardware or not.
-    default: 'default'
+      - When not set AWS will default to C(default).
     version_added: "2.4"
+    type: str
+    choices: ['default', 'dedicated']
+  associate_public_ip_address:
+    description:
+      - The I(wait_timeout) option does nothing and will be removed in Ansible 2.14.
```
Comment (jillr) [inline lib/ansible/modules/cloud/amazon/ec2_lc.py:147]:
```suggestion
      - The I(associate_public_ip_address) option does nothing and will be removed in Ansible 2.14.
```

**Review by gundalow — APPROVED** [review]

**Review by jillr — COMMENTED** [review]
Missed the merged, adding these comments so I don't lose them in an unsaved review while I'm working on it and will open a new PR for them.

**Pull request merged**